### PR TITLE
Implemented ddnss

### PR DESCRIPTION
**Pull request by seriouz**
libdns implementation for [ ddnss(.de)](https://ddnss.de) a german dns service.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+		log.Fatal(err)
+	}
+
+	bodyBytes, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
+
+	doc, err := htmlquery.Parse(strings.NewReader(string(bodyBytes)))
```
Comment (mholt) [inline client.go:220]: could probably use `bytes.NewReader()` instead

File: client.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+	nodes, err := htmlquery.QueryAll(doc, "//u")
+	if err != nil {
+		panic(`not a valid XPath expression.`)
+	}
+
+	domains := []string{}
+	for _, n := range nodes {
+		domain := strings.TrimSpace(n.FirstChild.Data)
+		log.Printf("[DDNSS] Found domain %s", domain)
```
Comment (mholt) [inline client.go:233]: Remove dev logs before publishing

File: provider.go [hunk 3]
```diff
@@ -1,47 +1,91 @@
...
-// when methods are called; sync.Once can help with this, and/or you can use a
-// sync.(RW)Mutex in your Provider struct to synchronize implicit provisioning.
-
-// Provider facilitates DNS record manipulation with <TODO: PROVIDER NAME>.
+// Provider facilitates DNS record manipulation with ddnss.
 type Provider struct {
-	// TODO: put config fields here (with snake_case json
-	// struct tags on exported fields), for example:
-	APIToken string `json:"api_token,omitempty"`
+	APIToken       string ``
```
Comment (mholt) [inline provider.go:14]: Fill out the JSON struct tags as the example.

File: client.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	}
+
+	return nil, fmt.Errorf("DDNSS request failed, expected (OK) but got url: [%s], body: %s", u, body)
+}
+
+func getMainDomain(domain string) string {
+	domain = strings.TrimSuffix(domain, ".")
+	log.Printf("getMainDomain start %s", domain)
+	domain = strings.TrimLeft(domain, "_acme-challenge.")
+	log.Printf("getMainDomain done %s", domain)
```
Comment (mholt) [inline client.go:146]: Remove all dev logs

File: client.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	ddnssDomainListUrl := "https://ddnss.de/ua/vhosts_list.php"
+	method := "POST"
+
+	options := cookiejar.Options{
+		PublicSuffixList: publicsuffix.List,
+	}
+
+	jar, err := cookiejar.New(&options)
+	if err != nil {
+		log.Fatal(err)
```
Comment (mholt) [inline client.go:164]: Code must not exit -- return errors instead

File: client.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	}
+
+	payload := &bytes.Buffer{}
+	writer := multipart.NewWriter(payload)
+	_ = writer.WriteField("action", "login")
+	_ = writer.WriteField("username", username)
+	_ = writer.WriteField("passwd", password)
+	err = writer.Close()
+	if err != nil {
+		fmt.Println(err)
```
Comment (mholt) [inline client.go:174]: Remove lines like this

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks; it will need some changes before it's ready to use though.

(Also, the copyright is in ddnss.de's name, is that correct? Or do you want it in your name?)

These are just my remarks from the first pass. Please be sure to remove all calls to Fatal() and log.Print*, fmt.Print*, etc.

File: client.go (hunk 1 above)
Comment (seriouz) [reply client.go]: Done

File: provider.go (hunk 3 above)
Comment (seriouz) [reply provider.go]: Like so?

**Comment by seriouz** [comment]
> 

I've done your requested changes. Updated the copyright, too

**Comment by seriouz** [comment]
Question: I've done the changes for [caddy-dns](https://github.com/caddy-dns) too. Where can i make the pull request for it?

**Comment by seriouz** [comment]
@mholt Question: I've done the changes for [caddy-dns](https://github.com/caddy-dns) too. Where can i make the pull request for it?

File: client.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+	body := string(bodyBytes)
+
+	doc, err := htmlquery.Parse(strings.NewReader(body))
+	if err != nil {
+		panic(`not a valid XPath expression.`)
+	}
+
+	nodes, err := htmlquery.QueryAll(doc, "//font")
+	if err != nil {
+		panic(`not a valid XPath expression.`)
```
Comment (mholt) [inline client.go:110]: Please don't panic, return errors instead. (Replace all panics with error returns)

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks, looks much better. One last thing I noticed after a quick skim is that this code panics; please remove panics and return errors instead.

File: client.go (hunk 7 above)
Comment (seriouz) [reply client.go]: Done! Using errors now.

**Review by mholt — APPROVED** [review]
LGTM enough to create the repo, so I'll invite you and you can push to that repo. Thanks for contributing!

**Comment by mholt** [comment]
(Are you using this with Caddy? If so I'll create a repo in [caddy-dns](https://github.com/caddy-dns) for you to make it a Caddy module as well!)

**Pull request closed without merging**